File: backend/app/schemas/tenant.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class TenantCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = Field(min_length=1, max_length=64)
    code: str | None = Field(default=None, max_length=32)

    @field_validator("name")
    @classmethod
    def strip_name(cls, value: str) -> str:
        name = value.strip()
        if not name:
            raise ValueError("企业名称不能为空")
        return name

    @field_validator("code")
    @classmethod
    def normalize_code(cls, value: str | None) -> str | None:
        if value is None:
            return None
        code = value.strip().lower()
        return code or None


class TenantOut(BaseModel):
    id: int
    name: str
    code: str
    status: str
    created_by: int
    created_at: datetime
    updated_at: datetime
    my_role: str
    my_status: str
    is_owner: bool

    model_config = {"from_attributes": True}


class TenantContextOut(BaseModel):
    user_id: int
    tenant_id: int
    member_id: int
    is_owner: bool
    role: str


class MemberCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    user_id: int = Field(gt=0)


class MemberUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: str = Field(min_length=1, max_length=16)

    @field_validator("status")
    @classmethod
    def normalize_status(cls, value: str) -> str:
        return value.strip().upper()


class MemberOut(BaseModel):
    id: int
    tenant_id: int
    user_id: int
    role: str
    status: str
    joined_at: datetime
    display_name: str
    email: str

File: packages/shared/src/types/tenant.ts
/** 与后端 `TenantOut` / `MemberOut` / `TenantContextOut` 对齐。 */

export interface Tenant {
  id: number;
  name: string;
  code: string;
  status: string;
  created_by: number;
  created_at: string;
  updated_at: string;
  my_role: string;
  my_status: string;
  is_owner: boolean;
}

export interface TenantContextInfo {
  user_id: number;
  tenant_id: number;
  member_id: number;
  is_owner: boolean;
  role: string;
}

export interface TenantMember {
  id: number;
  tenant_id: number;
  user_id: number;
  role: string;
  status: string;
  joined_at: string;
  display_name: string;
  email: string;
}

export interface TenantCreatePayload {
  name: string;
  code?: string | null;
}

export const TENANT_HEADER = 'X-Tenant-ID';

export const TENANT_STATUS = {
  active: 'ACTIVE',
  disabled: 'DISABLED',
} as const;

export const MEMBER_STATUS = {
  active: 'ACTIVE',
  disabled: 'DISABLED',
} as const;

export const TENANT_ERROR_CODE = {
  missingContext: 40030,
  forbidden: 40310,
  notFound: 40410,
} as const;
